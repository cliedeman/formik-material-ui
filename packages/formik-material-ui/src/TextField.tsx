import * as React from 'react';
import MuiTextField, { TextFieldProps as MuiTextFieldProps } from '@material-ui/core/TextField';
import { FieldProps, getIn } from 'formik';

export interface TextFieldProps
  extends FieldProps,
    Omit<MuiTextFieldProps, 'name' | 'value' | 'error'> {
  touchOnChange?: boolean
}

export function fieldToTextField({
  disabled,
  field,
  form: { isSubmitting, touched, errors, setTouched },
  touchOnChange,
  ...props
}: TextFieldProps): MuiTextFieldProps {
  const fieldError = getIn(errors, field.name);
  const showError = getIn(touched, field.name) && !!fieldError;

  return {
    ...props,
    ...field,
    onChange(e) {
      if (touchOnChange && !getIn(touched, field.name)) {
        setTouched({ [field.name]: true });
      }
      const onChange = props.onChange || field.onChange;
      onChange && onChange(e);
    },
    error: showError,
    helperText: showError ? fieldError : props.helperText,
    disabled: disabled ?? isSubmitting,
    variant: props.variant,
  };
}

export function TextField({ children, ...props }: TextFieldProps) {
  return <MuiTextField {...fieldToTextField(props)}>{children}</MuiTextField>;
}

TextField.displayName = 'FormikMaterialUITextField';
