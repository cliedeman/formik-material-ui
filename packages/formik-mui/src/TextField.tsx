import * as React from 'react';
import {
  type TextFieldProps as MuiTextFieldProps,
  TextField as MuiTextField
} from '@mui/material';
import { type FieldProps, getIn } from 'formik';

export interface TextFieldProps
  extends FieldProps,
    Omit<MuiTextFieldProps, 'name' | 'value' | 'error'> {}

export function fieldToTextField({
  disabled,
  field: { onBlur: fieldOnBlur, ...field },
  form: { isSubmitting, touched, errors },
  onBlur,
  helperText,
  ...props
}: TextFieldProps): MuiTextFieldProps {
  const fieldError = getIn(errors, field.name);
  const showError = getIn(touched, field.name) && !!fieldError;

  return {
    error: showError,
    helperText: showError ? fieldError : helperText,
    disabled: disabled ?? isSubmitting,
    onBlur:
      onBlur ??
      function (e) {
        fieldOnBlur(e ?? field.name);
      },
    ...field,
    ...props,
  };
}

export function TextField({ children, ...props }: TextFieldProps) {
  return <MuiTextField {...fieldToTextField(props)}>{children}</MuiTextField>;
}

TextField.displayName = 'FormikMaterialUITextField';
