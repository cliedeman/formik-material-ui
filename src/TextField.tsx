import * as React from 'react';
import MuiTextField, {
  TextFieldProps as MuiTextFieldProps,
} from '@material-ui/core/TextField';
import { FieldProps, getIn } from 'formik';
import { Omit } from './types';

export interface TextFieldProps
  extends FieldProps,
    Omit<MuiTextFieldProps, 'error' | 'name' | 'onChange' | 'value'> {}

export const fieldToTextField = ({
  field,
  form,
  disabled = false,
  ...props
}: TextFieldProps): MuiTextFieldProps => {
  const { name } = field;
  const { touched, errors, isSubmitting } = form;

  const fieldError = getIn(errors, name);
  const showError = getIn(touched, name) && !!fieldError;

  return {
    ...props,
    ...field,
    error: showError,
    helperText: showError ? fieldError : props.helperText,
    disabled: isSubmitting || disabled,
  };
};

export const TextField: React.ComponentType<TextFieldProps> = (
  props: TextFieldProps
) => <MuiTextField {...fieldToTextField(props)} />;

TextField.displayName = 'FormikMaterialUITextField';
