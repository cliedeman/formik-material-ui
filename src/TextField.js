/* @flow */
import TextField from '@material-ui/core/TextField';

import createComponent from './createComponent';

export default createComponent(
  TextField,
  ({ field, form: { touched, errors, isSubmitting }, disabled = false, ...props }) => {
    const { name } = field;

    return {
      ...props,
      ...field,
      error: touched[name] && !!errors[name],
      helperText: errors[name] ? errors[name] : props.helperText,
      disabled: isSubmitting || disabled,
    };
  }
);
