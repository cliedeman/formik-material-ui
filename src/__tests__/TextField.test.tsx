import * as React from 'react';
import { Formik, Field, Form } from 'formik';
import renderer from 'react-test-renderer';

import { TextField, TextFieldProps } from '../TextField';

test('TextField Renders Correctly', () => {
  const component = renderer.create(
    <Formik initialValues={{}} onSubmit={() => null}>
      <Form>
        <Field name="test" label="Text" component={TextField} />
      </Form>
    </Formik>
  );

  expect(component.toJSON()).toMatchSnapshot();
});

test('Wrapped TextField', () => {
  const TextFieldComponent = (p: TextFieldProps) => (
    <TextField variant="standard" {...p} />
  );

  const component = renderer.create(
    <Formik initialValues={{}} onSubmit={() => null}>
      <Form>
        <Field name="test" label="Text" component={TextFieldComponent} />
      </Form>
    </Formik>
  );

  expect(component.toJSON()).toMatchSnapshot();
});
