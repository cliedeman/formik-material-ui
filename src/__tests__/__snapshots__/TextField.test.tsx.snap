// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField Renders Correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="MuiFormControl-root-1"
  >
    <label
      className="MuiFormLabel-root-12 MuiInputLabel-root-5 MuiInputLabel-formControl-6 MuiInputLabel-animated-9"
      data-shrink={false}
    >
      Text
    </label>
    <div
      className="MuiInputBase-root-32 MuiInput-root-19 MuiInput-underline-23 MuiInputBase-formControl-33 MuiInput-formControl-20"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        className="MuiInputBase-input-42 MuiInput-input-27"
        disabled={false}
        name="test"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
      />
    </div>
  </div>
</form>
`;
